backend/cfg/cfg_regalloc_validate.ml

```diff
@@ -0,0 +1,845 @@
+    (* The instruction was present before. *)
+    | Some old_instr, false ->
+      if instr.desc <> old_instr.Instruction.desc
+      then
+        Cfg_regalloc_utils.fatal "The instruction's no. %d desc was changed" id;
+      verify_reg_array ~id ~name:"argument" ~reg_arr:old_instr.Instruction.arg
+        ~loc_arr:instr.arg;
+      verify_reg_array ~id ~name:"result" ~reg_arr:old_instr.Instruction.res
+        ~loc_arr:instr.res;
+      ()
+    (* Added spill/reload that wasn't before. *)
+    | None, true -> ()
+    | Some _, true ->
+      Cfg_regalloc_utils.fatal
+        "Register allocation changed existing instruction no. %d into a \
+         register allocation specific instruction"
+        id
+    | None, false ->
+      Cfg_regalloc_utils.fatal
+        "Register allocation added non-regalloc specific instruction no. %d" id
+
+  let verify t cfg =
+    Cfg_regalloc_utils.postcondition cfg;
+    let seen_ids =
+      Hashtbl.create
+        (Hashtbl.length t.instructions + Hashtbl.length t.terminators)
+    in
+    Cfg_with_layout.iter_instructions cfg
+      ~instruction:(make_instruction_helper t (verify_instr ~seen_ids))
+      ~terminator:(make_terminator_helper t (verify_instr ~seen_ids));
+    Hashtbl.iter
+      (fun id instr ->
+        (* CR azewierzejew for azewierzejew: We would like to do an actual check
+           whether [Prologue] can be removed. *)
+        let can_be_removed =
+          match instr.Instruction.desc with Prologue -> true | _ -> false
+        in
+        if Hashtbl.find_opt seen_ids id |> Option.is_none && not can_be_removed
+        then
+          Cfg_regalloc_utils.fatal
+            "Instruction no. %d was deleted by register allocator" id)
+      t.instructions;
+    Hashtbl.iter
+      (fun id _ ->
+        if Hashtbl.find_opt seen_ids id |> Option.is_none
+        then
+          Cfg_regalloc_utils.fatal
+            "Terminator no. %d was deleted by register allocator" id)
+      t.terminators
+end
+
+module Equation_set : sig
+  type t
+
+  val empty : t
+
+  val equal : t -> t -> bool
+
+  val union : t -> t -> t
+
+  val remove_result :
+    reg_res:Register.t array ->
+    loc_res:Location.t array ->
+    t ->
+    (t, string) Result.t
+
+  val verify_destroyed_locations :
+    destroyed:Location.t array -> t -> (unit, string) Result.t
+
+  val add_argument :
+    reg_arg:Register.t array -> loc_arg:Location.t array -> t -> t
+
+  val rename_loc : arg:Location.t -> res:Location.t -> t -> t
+
+  val rename_reg : arg:Register.t -> res:Register.t -> t -> t
+
+  val is_empty : t -> bool
+
+  val print : Format.formatter -> t -> unit
+end = struct
+  module Equation = struct
+    type t = Register.t * Location.t
+
+    let compare (r1, l1) (r2, l2) =
+      let r_cmp = Register.compare r1 r2 in
+      if r_cmp <> 0 then r_cmp else Location.compare l1 l2
+
+    let print ppf (r, l) =
+      Format.fprintf ppf "%a=%a" Register.print r Location.print l
+  end
+
+  include Set.Make (Equation)
```

Comment:
Is polymorphic compare slower than by-hand implementation could be?

Reply:
I think it usually is, because it involves a C call.
But do not take my word for it: it may be the case
that it does not make a material different for this
code. If you measure it, leave a comment with
your conclusion; if not, leave a CR-someday.

Reply:
I left CR-someday's